File: source-code/valuations/two-stage-dividend-discount-model.js
// +------------------------------------------------------------+
//   Model: Two-Stage Dividend Discount Model
//   © Copyright: https://discountingcashflows.com
// +------------------------------------------------------------+

Input(
  {
    _DISCOUNT_RATE: '',
    HIGH_GROWTH_YEARS: 5,
    HISTORICAL_YEARS: '',
    _HIGH_GROWTH_RATE: '',
    _HIGH_GROWTH_PAYOUT: '',
    _STABLE_GROWTH_IN_PERPETUITY: '',
    _STABLE_PAYOUT: '',
    BETA:'',
    _RISK_FREE_RATE: '',
    _MARKET_PREMIUM: '',
  },
  [{
    parent:'_DISCOUNT_RATE',
    children:['BETA', '_RISK_FREE_RATE', '_MARKET_PREMIUM']
  }]
);  

$.when(
  get_income_statement(),
  get_income_statement_ltm(),
  get_balance_sheet_statement(),
  get_balance_sheet_statement_quarterly(),
  get_cash_flow_statement(),
  get_cash_flow_statement_ltm(),
  get_profile(),
  get_dividends_annual(),
  get_prices_annual(),
  get_treasury(),
  get_fx(),
  get_risk_premium()).done(
  function(_income, _income_ltm, _balance, _balance_quarterly, _flows, _flows_ltm, _profile, _dividends, _prices, _treasury, _fx, _risk_premium){
  try{
    var response = new Response({
      income: _income,
      income_ltm: _income_ltm,
      balance: _balance,
      balance_quarterly: _balance_quarterly,
      balance_ltm: 'balance_quarterly:0',
      flows: _flows,
      flows_ltm: _flows_ltm,
      profile: _profile,
      treasury: _treasury,
      dividends: _dividends,
      prices: _prices,
      risk_premium: _risk_premium,
    }).toOneCurrency('income', _fx).merge('_ltm');
    response.balance[0]['date'] = response.prices[0]['date'] = 'LTM';
    // +---------------- ASSUMPTIONS SECTION -----------------+ 
    // Count the dividends. If there are no dividends, display a warning.
    var dividendsCount = response.dividends.length - 1;
    if(dividendsCount <= 0){
      throwError("The company does not currently pay dividends!");
      return;
    }
    if(dividendsCount > 10){
    	setAssumption('HISTORICAL_YEARS', 10);
    }
    else{
      // Set the default historical years to the number of historical dividends
      setAssumption('HISTORICAL_YEARS', dividendsCount);
    }
    // Set the stable growth in perpetuity to the 10 year treasury note
    setAssumption('_STABLE_GROWTH_IN_PERPETUITY', response.treasury.year10);
	
    // Set beta 
    if(response.profile.beta){
    	setAssumption('BETA', response.profile.beta);
    }
    else{
    	setAssumption('BETA', 1);
    }
    // Risk free rate is the yield of the 10 year treasury note
	setAssumption('_RISK_FREE_RATE', response.treasury.year10);
    setAssumption('_MARKET_PREMIUM', response.risk_premium.totalEquityRiskPremium );
    // Discount Rate is the cost of equity
    setAssumption('_DISCOUNT_RATE', toP(getAssumption('_RISK_FREE_RATE') + getAssumption('BETA') * getAssumption('_MARKET_PREMIUM')));
    
    var sensitivity = 0.01;
    var prefDividendsRatio = absolute((response.income[0].eps * response.income[0].weightedAverageShsOut - response.income[0].netIncome) / response.income[0].netIncome);
    var commonIncome = [];
    if( prefDividendsRatio > sensitivity ){
      commonIncome = ['eps:0', '*', 'weightedAverageShsOut:0'];
    }
    else{
      commonIncome = ['netIncome:0'];
    }
    // Setup Original Data
    var original_data = new DateValueData({
      'netIncome': new DateValueList(response.income, 'netIncome'),
      'totalStockholdersEquity': new DateValueList(response.balance, 'totalStockholdersEquity'),
      'weightedAverageShsOut': new DateValueList(response.income, 'weightedAverageShsOut'),
      'eps': new DateValueList(response.income, 'eps'),
      'adjDividend': new DateValueList(response.dividends, 'adjDividend'),
      'marketPrice': new DateValueList(response.prices, 'close'),
    });
    var currentDate = original_data.lastDate();
    var nextYear = currentDate + 1;
    var forecastEndDate = currentDate + getAssumption('HIGH_GROWTH_YEARS');
    
    // Compute historical values and ratios
    var historical_computed_data = original_data.setFormula({
      'commonIncome': commonIncome,
      'preferredStockDividends': ['netIncome:0', '-', 'commonIncome:0'],
      'dividendsPaidToCommon': ['adjDividend:0', '*', 'weightedAverageShsOut:0'],
      'bookValue': ['totalStockholdersEquity:0', '/', 'weightedAverageShsOut:0'],
      '_returnOnEquity': ['commonIncome:0', '/', 'totalStockholdersEquity:-1'],
      '_payoutRatio': ['adjDividend:0', '/', 'eps:0'],
      'retainedEarnings': ['eps:0', '-', 'adjDividend:0'],
      '_dividendYield': ['adjDividend:0', '/', 'marketPrice:0'],
      '_adjDividendGrowth': ['function:growth_rate', 'adjDividend'],
      'discountedAdjDividend': ['adjDividend:0'],
    }).compute();
    
    var averageReturnOnEquity = historical_computed_data.get('_returnOnEquity').sublist(nextYear - getAssumption('HISTORICAL_YEARS')).average();
    setAssumption('_STABLE_PAYOUT', toP(1-getAssumption('_STABLE_GROWTH_IN_PERPETUITY')/averageReturnOnEquity));
    
    var averagePayoutRatio = historical_computed_data.get('_payoutRatio').sublist(nextYear - getAssumption('HISTORICAL_YEARS')).average();
    // If the payout is higher than 100%, set it equal to the stable payout
    if(averagePayoutRatio > 1){
      setAssumption('_HIGH_GROWTH_PAYOUT', toP(getAssumption('_STABLE_PAYOUT')));
    }
    else{
      setAssumption('_HIGH_GROWTH_PAYOUT', toP(averagePayoutRatio));
    }
    // averageDividendGrowthRate stores the Historical Dividend Growth Rate
    // TODO: Remove '+ 1' from nextYear + 1
    var averageDividendGrowthRate = historical_computed_data.get('_adjDividendGrowth').sublist(nextYear - getAssumption('HISTORICAL_YEARS')).average();
    // Set the eps high growth rate equal to the historical dividend growth rate
    setAssumption('_HIGH_GROWTH_RATE', toP(averageDividendGrowthRate));
    
    // Compute 5 years of forecasted values and ratios
    var forecasted_data = historical_computed_data.removeDate('LTM').setFormula({
      'linearRegressionEps': ['function:linear_regression', 'eps', {slope: 1, start_date: nextYear - getAssumption('HISTORICAL_YEARS')}],
      'eps': ['function:compound', 'eps:start_date', {rate: getAssumption('_HIGH_GROWTH_RATE'), start_date: currentDate}],
      'adjDividend': ['eps:0', '*', getAssumption('_HIGH_GROWTH_PAYOUT')],
      'discountedAdjDividend': ['function:discount', 'adjDividend', {rate: getAssumption('_DISCOUNT_RATE'), start_date: currentDate}],
      '_adjDividendGrowth': ['function:growth_rate', 'adjDividend'],
    }).setEditable(_edit(), {
      start_date: nextYear,
      keys: ['eps', 'adjDividend'],
    }).compute({'forecast_end_date': forecastEndDate});
    // +------------- END OF ASSUMPTIONS SECTION -------------+
    
    // +---------------- MODEL VALUES SECTION ----------------+
    // Discount the projected dividends and sum them
    var sumOfDiscountedDividends = forecasted_data.get('discountedAdjDividend').sublist(nextYear).sum();
    // Calculate the discounted terminal value
    var stableEps = forecasted_data.get('eps').lastValue() * (1 + getAssumption('_STABLE_GROWTH_IN_PERPETUITY'));
    var stableDividend = stableEps * getAssumption('_STABLE_PAYOUT');
    var terminalValue = stableDividend/(getAssumption('_DISCOUNT_RATE') - getAssumption('_STABLE_GROWTH_IN_PERPETUITY')); 
    var discountedTerminalValue = terminalValue/Math.pow(1 + getAssumption('_DISCOUNT_RATE'), getAssumption('HIGH_GROWTH_YEARS'));
    
    // The final value calculated by the Two-Stage Dividend Discount Model
    var valueOfStock = discountedTerminalValue + sumOfDiscountedDividends;
    
    var currency = response.currency;
    // If we are calculating the value per share for a watch, we can stop right here.
    if(_StopIfWatch(valueOfStock, currency)){
      return;
    }
    _SetEstimatedValue(valueOfStock, currency);
    print(sumOfDiscountedDividends, "Sum of discounted dividends", '#', currency);
    print(discountedTerminalValue, "Discounted terminal value", '#', currency);
    print(terminalValue, "Terminal value", '#', currency);
    print(stableDividend, "Dividend in stable phase", '#', currency);
    print(stableEps, "Eps in stable phase", '#', currency);
    print(averageDividendGrowthRate, "Average historical Dividend Growth Rate", '%');
    print(averagePayoutRatio, "Average historical Payout Ratio", '%');
    print(averageReturnOnEquity, "Average historical Return on Equity", '%');
    // +------------- END OF MODEL VALUES SECTION ------------+
    
    // +------------------- CHARTS SECTION -------------------+
    forecasted_data.renderChart({
      start_date: nextYear - getAssumption('HISTORICAL_YEARS'),
      keys: ['eps', 'adjDividend', 'linearRegressionEps', 'discountedAdjDividend'],
      properties: {
        title: 'Historical and Projected Dividends',
        currency: currency,
        disabled_keys: ['linearRegressionEps', 'discountedAdjDividend'],
      }
    });
    // +------------------- CHARTS SECTION -------------------+
    
    // +------------------- TABLES SECTION -------------------+
    // Dividend Table
    forecasted_data.renderTable({
      start_date: currentDate,
      keys: ['eps', 'adjDividend', '_adjDividendGrowth', 'discountedAdjDividend'],
      rows: ['EPS', 'Dividends', '{%} Dividend Growth Rate', 'Discounted Dividend'],
      properties: {
        'title': 'Projected data',
        'currency': currency,
      },
    });
    
    // Historical Table
    historical_computed_data.renderTable({
      start_date: nextYear - getAssumption('HISTORICAL_YEARS'),
      keys: ['netIncome', 'totalStockholdersEquity', '_returnOnEquity', 'dividendsPaidToCommon',
             '_payoutRatio', 'weightedAverageShsOut', 'marketPrice', 'eps', 'adjDividend', '_adjDividendGrowth', '_dividendYield'],
      rows: ['Net income', 'Equity', '{%} Return on equity', 'Dividends paid', 
             '{%} Payout ratio', 'Shares outstanding', '{PerShare} Reference market price', '{PerShare} EPS',
             '{PerShare} Dividends', '{%} Dividend Growth Rate', '{%} Dividend yield'],
      properties: {
        'title': 'Historical data',
        'currency': currency,
        'number_format': 'M',
        'display_averages': true,
        'column_order': 'descending',
      },
    });
    // +---------------- END OF TABLES SECTION ---------------+
  }
  catch (error) {
    throwError(error);
  }
});

Description(`
	<h5>Two-Stage Dividend Discount Model</h5>
	<p>Used to estimate the value of companies based on two stages of growth. An initial period of high growth, calculated using <b>[Sum of Discounted Dividends]</b>, followed by a period of stable growth, calculated using <b>[Discounted Terminal Value]</b>.</p>
	<p class='text-center'>Read more: <a href='https://github.com/DiscountingCashFlows/Documentation/blob/main/models-documentation/dividend-discount-models.md#two-stage-dividend-discount-model-source-code' target='_blank'><i class="fab fa-github"></i> GitHub Documentation</a></p>
`);

File: source-code/valuations/two-stage-dividend-discount-model.py
"""
    Model: Two-Stage Dividend Discount Model

    © Copyright discountingcashflows.com
"""

# Render model description
model.render_description(r"""
## Two-Stage Dividend Discount Model

Used to estimate the value of companies based on two stages of growth. An initial period of high growth, calculated using **[Sum of Discounted Dividends]**, followed by a period of stable growth, calculated using **[Discounted Terminal Value]**.

Read more: [GitHub Documentation](https://github.com/DiscountingCashFlows/Documentation/blob/main/models-documentation/dividend-discount-models.md#two-stage-dividend-discount-model-source-code)
""")

# Initialize assumptions
assumptions.init({
    "data": {
        "%discount_rate": None,
        "beta": data.get("profile:beta", default=1),
        "%risk_free_rate": data.get("treasury:year10"),
        "%market_premium": data.get("risk:totalEquityRiskPremium"),
        "high_growth_years": 5,
        "historical_years": None,
        "%high_growth_rate": None,
        "%high_growth_payout": None,
        "%stable_growth_in_perpetuity": data.get("treasury:year10"),
        "%stable_payout": None,
    },
    "hierarchies": [{
        "parent": "%discount_rate",
        "children": ["beta", "%risk_free_rate", "%market_premium"]
    }]
})

# Set the discount rate to the cost of equity
risk_free_rate = assumptions.get("%risk_free_rate")
beta = assumptions.get("beta")
market_premium = assumptions.get("%market_premium")
cost_of_equity = risk_free_rate + beta * market_premium

assumptions.set("%discount_rate", cost_of_equity)

# Count all dividends excluding None and 0 values
dividend_count = data.count("dividend:adjDividend:*", properties={"except_values": [None, 0]})

if dividend_count <= 0:
    raise ValueError("The company does not currently pay dividends!")

elif dividend_count > 10:
    assumptions.set("historical_years", 10)

else:
    # Set the default historical years to the number of historical dividends
    assumptions.set("historical_years", dividend_count)

# Compute historical values and ratios
data.compute({
    "dividendsPaidToCommon": "dividend:adjDividend * income:weightedAverageShsOut",
    "%returnOnEquity": "income:netIncome / balance:totalStockholdersEquity:-1",
    "%payoutRatio": "dividend:adjDividend / income:eps",
    "%dividendYield": "dividend:adjDividend / quote:close",
    "%adjDividendGrowth": "function:growth:dividend:adjDividend",
    "#discountedAdjDividend": "dividend:adjDividend",
})

# Calculate average return on equity
average_return_on_equity = data.average(f"%returnOnEquity:{-assumptions.get('historical_years')}->0")
stable_payout = 1 - assumptions.get("%stable_growth_in_perpetuity") / average_return_on_equity
assumptions.set("%stable_payout", stable_payout)

# Calculate the average payout ratio
average_payout_ratio = data.average(f"%payoutRatio:{-assumptions.get('historical_years')}->0")
if average_payout_ratio > 1:
    assumptions.set("%high_growth_payout", assumptions.get("%stable_payout"))
else:
    assumptions.set("%high_growth_payout", average_payout_ratio)

# Calculate average dividend growth rate
average_dividend_growth_rate = data.average(f"%adjDividendGrowth:{-assumptions.get('historical_years')}->0")
assumptions.set("%high_growth_rate", average_dividend_growth_rate)

data.compute({
    "#linearRegressionEps": f"function:linear_regression:income:eps start:{-assumptions.get('historical_years')}"
}, forecast=assumptions.get("high_growth_years"))

# Compute 5 years of forecasted values and ratios
data.compute({
    "income:eps": f"function:compound:{data.get('#linearRegressionEps')} rate:{assumptions.get('%high_growth_rate')} offset:-1",
    "dividend:adjDividend": f"income:eps * {assumptions.get('%high_growth_payout')}",
    "#discountedAdjDividend": f"function:discount:dividend:adjDividend rate:{assumptions.get('%discount_rate')}",
    "%adjDividendGrowth": "function:growth:dividend:adjDividend",
}, forecast=assumptions.get("high_growth_years"))

# Discount the projected dividends and sum them
sum_of_discounted_dividends = data.sum(f"#discountedAdjDividend:1->{assumptions.get('high_growth_years')}")

# Calculate the discounted terminal value
stable_eps = data.get(f"income:eps:{assumptions.get('high_growth_years')}") * (1 + assumptions.get("%stable_growth_in_perpetuity"))
stable_dividend = stable_eps * assumptions.get("%stable_payout")
terminal_value = stable_dividend / (assumptions.get("%discount_rate") - assumptions.get("%stable_growth_in_perpetuity"))
discounted_terminal_value = terminal_value / ((1 + assumptions.get("%discount_rate")) ** assumptions.get("high_growth_years"))

# The final value calculated by the Two-Stage Dividend Discount Model
final_value = discounted_terminal_value + sum_of_discounted_dividends

model.set_final_value({
    "value": final_value,
    "units": "$"
})

# Render results
model.render_results([
    [sum_of_discounted_dividends, "Sum of discounted dividends", "$"],
    [discounted_terminal_value, "Discounted terminal value", "$"],
    [terminal_value, "Terminal value", "$"],
    [stable_dividend, "Dividend in stable phase", "$"],
    [stable_eps, "EPS in stable phase", "$"],
    [average_dividend_growth_rate, "Average historical Dividend Growth Rate", "%"],
    [average_payout_ratio, "Average historical Payout Ratio", "%"],
    [average_return_on_equity, "Average historical Return on Equity", "%"],
])

# Render chart for projected dividends
model.render_chart({
    "data": {
        "income:eps": "Earnings Per Share",
        "dividend:adjDividend": "Dividend",
        "#linearRegressionEps": "Linear Regression of EPS",
        "#discountedAdjDividend": "Discounted Dividends"
    },
    "start": -assumptions.get('historical_years'),
    "properties": {
        "title": "Historical and Projected Dividends",
        "set_editable": [
            "income:eps",
            "dividend:adjDividend"
        ],
        "hidden_keys": [
            "#linearRegressionEps",
            "#discountedAdjDividend"
        ],
    }
})

# Render Dividend Table
model.render_table({
    "data": {
        "income:eps": "Earnings Per Share",
        "dividend:adjDividend": "Dividend per Share",
        "%adjDividendGrowth": "Dividend Growth Rate",
        "#discountedAdjDividend": "Discounted Dividend",
    },
    "start": -1,
    "properties": {
        "title": "Projected data",
        "column_order": "ascending",
    },
})

# Render Historical Table
model.render_table({
    "data": {
        "income:netIncome": "Net Income",
        "balance:totalStockholdersEquity": "Equity",
        "%returnOnEquity": "Return on Equity",
        "dividendsPaidToCommon": "Common Dividends",
        "%payoutRatio": "Payout Ratio",
        "income:weightedAverageShsOut": "Shares Outstanding",
        "quote:close": "Reference Market Price",
        "income:eps": "EPS",
        "dividend:adjDividend": "Dividend per Share",
        "%adjDividendGrowth": "Dividend Growth Rate",
        "%dividendYield": "Dividend Yield",
    },
    "start": -assumptions.get('historical_years'),
    "end": 0,
    "properties": {
        "title": "Historical data",
        "number_format": "M",
        "display_averages": True
    },
})

assumptions.set_description({
    "%discount_rate": r"""
        ## Discount Rate

        $
        \text{Discount Rate} = \text{Cost of Equity}
          = \text{Risk Free Rate} + \text{Beta} \times \text{Market Premium}
        $

        The cost of equity is the theoretical rate of return that an equity investment should generate. It is calculated using the CAPM formula.

        [Read More](https://www.investopedia.com/terms/c/costofequity.asp#mntl-sc-block_1-0-20)
    """,

    "beta": r"""
        ## Beta

        Beta is a value that measures the price fluctuations (volatility) of a stock with respect to fluctuations in the overall stock market.
    """,

    "%risk_free_rate": r"""
        ## Risk-Free Rate

        The risk-free rate represents the interest an investor would expect from an absolutely risk-free investment over a specified period of time.
        By default, it is equal to the current yield of the U.S. 10 Year Treasury Bond.
    """,

    "%market_premium": r"""
        ## Market Premium

        Market risk premium represents the excess returns over the risk-free rate that investors expect for taking on the incremental risks connected to the equities market.
    """,

    "high_growth_years": r"""
        ## High Growth Years

        The estimated number of years during which the company is expected to experience high growth. After this period, the company will transition to a stable growth phase.
    """,

    "historical_years": r"""
        ## Historical Years

        The number of historical years used to calculate averages for historical data.
    """,

    "%high_growth_rate": r"""
        ## High Growth Rate

        The estimated growth rate of Earnings Per Share (EPS) during the high-growth phase.
    """,

    "%high_growth_payout": r"""
        ## High Growth Payout

        The estimated dividend payout rate expressed as a percentage of Earnings Per Share (EPS) during the high-growth phase.
    """,

    "%stable_growth_in_perpetuity": r"""
        ## Stable Growth in Perpetuity

        The estimated stable rate at which the company's Earnings Per Share (EPS) will grow in perpetuity after the last projected period.
    """,

    "%stable_payout": r"""
        ## Stable Payout

        The estimated dividend payout rate expressed as a percentage of Earnings Per Share (EPS) after the last projected period.
    """,
})

print("End of model")
